tests/mcproto/types/test_nbt.py

```diff
@@ -19,978 +19,380 @@
 
-    comp2 = CompoundNBT(
-        [ByteNBT(0, name="test"), ByteNBT(1, name="test2"), ByteNBT(2, name="test3")],
-        "comp2",
-    )
+    comp2 = CompoundNBT([ByteNBT(0, name="test"), ByteNBT(1, name="test2"), ByteNBT(2, name="test3")], "comp2")
     assert comp1 != comp2
 
     assert comp1 != ByteNBT(0, name="comp")
 
 
 # endregion
-# region IntArrayNBT
-
-
-@pytest.mark.parametrize(
-    ("payload", "error"),
-    [
-        ([0, "test"], ValueError),
-        ([0, None], ValueError),
-        ([0, 1 << 31], OverflowError),
-        ([0, -(1 << 31) - 1], OverflowError),
-    ],
-)
-def test_serialize_intarray_fail(payload: PayloadType, error: type[Exception]):
-    """Test serialization of NBT INTARRAY tag with invalid value."""
-    with pytest.raises(error):
-        IntArrayNBT(payload, "test").serialize()
-
-
-def test_deserialize_intarray_fail():
-    """Test deserialization of NBT INTARRAY tag with invalid value."""
-    # Not enough data for 1 element
-    buffer = Buffer(bytearray([0x0B, 0, 0, 0, 1, 0, 0, 0]))
-    with pytest.raises(IOError):
-        IntArrayNBT.deserialize(buffer, with_name=False)
-
-    # Not enough data for the size
-    buffer = Buffer(bytearray([0x0B, 0, 0, 0]))
-    with pytest.raises(IOError):
-        IntArrayNBT.read_from(buffer, with_name=False)
-
-    # Not enough data to start the 2nd element
-    buffer = Buffer(bytearray([0x0B, 0, 0, 0, 2, 1, 0, 0, 0]))
-    with pytest.raises(IOError):
-        IntArrayNBT.read_from(buffer, with_name=False)
-
-
-# endregion
-# region LongArrayNBT
-
-
-@pytest.mark.parametrize(
-    ("payload", "error"),
-    [
-        ([0, "test"], ValueError),
-        ([0, None], ValueError),
-        ([0, 1 << 63], OverflowError),
-        ([0, -(1 << 63) - 1], OverflowError),
-    ],
-)
-def test_serialize_deserialize_longarray_fail(payload: PayloadType, error: type[Exception]):
-    """Test serialization/deserialization of NBT LONGARRAY tag with invalid value."""
-    with pytest.raises(error):
-        LongArrayNBT(payload, "test").serialize()
-
 
-def test_deserialize_longarray_fail():
-    """Test deserialization of NBT LONGARRAY tag with invalid value."""
-    # Not enough data for 1 element
-    buffer = Buffer(bytearray([0x0C, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0]))
-    with pytest.raises(IOError):
-        LongArrayNBT.deserialize(buffer, with_name=False)
+# region ListNBT
 
-    # Not enough data for the size
-    buffer = Buffer(bytearray([0x0C, 0, 0, 0]))
-    with pytest.raises(IOError):
-        LongArrayNBT.read_from(buffer, with_name=False)
 
-    # Not enough data to start the 2nd element
-    buffer = Buffer(bytearray([0x0C, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0]))
-    with pytest.raises(IOError):
-        LongArrayNBT.read_from(buffer, with_name=False)
+def test_intarray_negative_length():
+    """Test IntArray with negative length."""
+    buffer = Buffer(b"\x0b\x00\x01a\xff\xff\xff\xff")
+    assert IntArrayNBT.read_from(buffer) == IntArrayNBT([], "a")
```

Comment:
Really? This doesn't fail if we get negative length? Is that in the spec?

Reply:
It's surprising but I found this on https://wiki.vg/NBT which is what I based my implementation on (and it seems to be up to date)

> A list of nameless tags, all of the same type. The list is prefixed with the Type ID of the items it contains (thus 1 byte), and the length of the list as a signed integer (a further 4 bytes). If the length of the list is 0 or negative, the type may be 0 (TAG_End) but otherwise it must be any other type. (The notchian implementation uses TAG_End in that situation, but another reference implementation by Mojang uses 1 instead; **parsers should accept any type if the length is <= 0**).

(Emphasis mine)